lalamo/modules/classifier.py

```diff
@@ -0,0 +1,306 @@
+from collections.abc import Mapping
+from dataclasses import dataclass, replace
+from typing import Self
+
+import equinox as eqx
+import jax
+from jax import vmap
+from jaxtyping import Array, DTypeLike, Float, Int, PRNGKeyArray
+
+from lalamo.common import ParameterTree
+from lalamo.modules.normalization import NormalizationConfig
+from lalamo.modules.transformer import Normalization, Transformer, TransformerConfig, TransformerForwardPassConfig
+
+from .activations import GELU, Activation, SiLU
+from .common import ForwardPassMode, LalamoModule
+from .embedding import EmbeddingBase, EmbeddingConfig
+from .linear import LinearBase, LinearConfig
+from .rope import PositionalEmbeddings
+from .transformer_layer import TransformerLayerResult
+from .utils import vmap_twice
+
+__all__ = [
+    "Classifier",
+    "ClassifierActivationTrace",
+    "ClassifierConfig",
+    "ClassifierResult",
+]
+
+def activation_from_str(activation: str) -> Activation:
+    supported_activations = {
+        "silu" : SiLU,
+        "gelu" : GELU,
+    }
+    if activation in supported_activations:
+        return supported_activations[activation]
+
+    raise ValueError(
+        f"Only activations from the following list are supported by Classifier: {supported_activations.keys()}"
+    )
+
+@dataclass(frozen=True)
+class PredictionHeadConfig:
+    dense_config: LinearConfig
+    activation: Activation
+    normalization_config: NormalizationConfig
+    use_dense_bias: bool
+    use_norm_bias: bool #NOTE: currently not used because Normalization class does not support bias
+
+    def empty(self, input_size:int, output_size: int) -> "PredictionHead":
+        dense_layer = self.dense_config.empty(
+            input_dim=input_size,
+            output_dims=(output_size,),
+            has_biases=self.use_dense_bias,
+        )
+        activation = self.activation
+        norm = self.normalization_config.empty(output_size)
+
+        return PredictionHead(config=self, dense=dense_layer, activation=activation, norm=norm)
+
+    def random_init(self, input_size:int, output_size: int, key: PRNGKeyArray) -> "PredictionHead":
+        dense_key, _ = jax.random.split(key)
+        dense_layer = self.dense_config.random_init(
+            input_size,
+            (output_size,),
+            has_biases=self.use_dense_bias,
+            key=dense_key)
+        activation = self.activation
+        norm = self.normalization_config.empty(output_size)
+
+        return PredictionHead(config=self, dense=dense_layer, activation=activation, norm=norm)
+
+
+class PredictionHead(LalamoModule[PredictionHeadConfig]):
+    dense: LinearBase
+    activation: Activation
+    norm: Normalization
+
+    def __call__(self, inner_features: Float[Array, " in_channels"])->Float[Array, " out_channels"]:
+        dense_outs = vmap_twice(self.dense)(inner_features)
```

Comment:
Instead of calling vmap_twice every line just create a separate function call_unbatched and wrap call_unbatched in vmap_twice